keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Many workloads benefit significantly from being aware of the topology of the node they are scheduled on.
+However, Kubernetes currently lacks a built-in mechanism to directly expose this node topology information to Pods.
+This KEP proposes a built-in admission plugin to copy standard node topology labels to Pods.
+Once copied, these labels can be consumed via the Downward API, just like any other Pod label.
+This approach eliminates the need for workarounds, such as custom init containers with elevated privileges,
+promoting a more secure and consistent solution.
+
+## Motivation
+
+Topology awareness is crucial for a growing number of applications and workloads on Kubernetes.
+Knowing the Pods's location within the cluster's topology allows for significant performance
+optimizations and improved resilience. End-user feedback has highlighted the following key use cases:
+* High-bandwidth AI/ML workloads, especially those employing distributed training or inference,
+  demonstrate a substantial need for topology awareness. The performance of these workloads is
+  highly sensitive to communication latency between GPUs. By preferentially scheduling Pods
+  requiring GPU-to-GPU communication within the same zone or rack, training and inference
+  times can be dramatically reduced.  Importantly, while Kubernetes can facilitate this topology-aware
+  placement, the actual orchestration of optimized GPU-to-GPU communication often relies on topology awareness
+  within the machine learning framework itself (e.g., Ray, PyTorch Distributed. etc).  Kubernetes provides the
+  necessary foundation for these frameworks to operate efficiently, but the frameworks are ultimately responsible
+  for leveraging that foundation.
+* CNI plugins can leverage topology information to establish more efficient network paths,
+  reducing latency and increasing throughput. For instance, a CNI could prioritize connections
+  within the same availability zone or rack.
+* Distributed stateful applications, such as sharded databases, can use topology information to improve fault tolerance.
+  By spreading replicas and traffic across different failure domains (e.g., zones, racks), these applications can achieve
+  higher availability and resilience to failures in any one topology.
+
+Today, applications obtain topology information by using an init container to query the Kubernetes API for the underlying node,
+extracting the necessary data from the Node resource. As more workloads and use cases benefit from topology awareness,
+we aim to simplify access to this information for Pods via the Downward API.
+
+### Goals
+
+* Values from Node labels `topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname` are made
+  available via downward API
+* Additional node labels can be made available via downward API using admission webhooks that mutate `pods/binding`.
```

Comment:
can you elaborate on this? Is it covered somewhere in the KEP on how to implement this admission webhook?

Reply:
The implementation of the mutating webhook itself will be very similar to the built-in admission plugin we are going to write, except it's an external mutating webhook (requiring a MutatingWebhookConfiguration) and the copied labels can be customized.

Reply:
The details I cannot find is what this webhook should copy and where. And what will happen automatically after this and how it may conflict with the built-in functionality